onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc

```diff
@@ -32,81 +32,103 @@ ONNX_OPERATOR_KERNEL_EX(
-  sh.MainFunctionBody() << "  let packed_qkv_indices = " << packed_qkv.OffsetToIndices("global_idx") << ";\n"
-                        << "  let input_data = " << packed_qkv.GetByOffset("global_idx") << ";\n"
-                        << "  let index = " << packed_qkv.IndicesGet("packed_qkv_indices", "2") << ";\n"
-                        << "  if (index < uniforms.hidden_size) {\n"
-                        << "    " << query.SetByIndices("packed_qkv_indices", "input_data") << ";\n"
-                        << "  } else if (index < (uniforms.hidden_size + uniforms.kv_hidden_size)) {\n"
-                        << "    var key_indices = packed_qkv_indices;\n"
-                        << "   " << key.IndicesSet("key_indices", "2", "u32(index - uniforms.hidden_size)") << ";\n"
-                        << "   " << key.SetByIndices("key_indices", "input_data") << ";\n"
-                        << "  } else {\n"
-                        << "    var val_indices = packed_qkv_indices;\n"
-                        << "   " << value.IndicesSet("val_indices", "2", "u32(index - uniforms.hidden_size - uniforms.kv_hidden_size)") << ";\n"
-                        << "   " << value.SetByIndices("val_indices", "input_data") << ";\n"
-                        << "  }";
+  const auto& packed_qkv = sh.AddInput("packed_qkv", ShaderUsage::UseUniform);
+  const auto& query = sh.AddOutput("query", ShaderUsage::UseUniform);
+  const auto& key = sh.AddOutput("key", ShaderUsage::UseUniform);
+  const auto& value = sh.AddOutput("val", ShaderUsage::UseUniform);
+
+  sh.MainFunctionBody() << "  // Dispatch over Q domain only\n"
+                        << "  let q_flat_idx = global_idx;\n"
+                        << "  let batch_idx = q_flat_idx / (uniforms.sequence_length * uniforms.hidden_size);\n"
+                        << "  let remainder = q_flat_idx % (uniforms.sequence_length * uniforms.hidden_size);\n"
+                        << "  let seq_idx = remainder / uniforms.hidden_size;\n"
+                        << "  let hidden_idx = remainder % uniforms.hidden_size;\n"
+                        << "\n"
+                        << "  // Calculate base offset for this (batch, seq) position in packed_qkv\n"
+                        << "  // Layout per token: [Q(hidden_size), K(kv_hidden_size), V(kv_hidden_size)]\n"
+                        << "  let token_size = uniforms.hidden_size + 2u * uniforms.kv_hidden_size;\n"
+                        << "  let base_offset = batch_idx * uniforms.sequence_length * token_size + seq_idx * token_size;\n"
+                        << "\n"
+                        << "  // Read and write Q\n"
+                        << "  let q_offset = base_offset + hidden_idx;\n"
+                        << "  let q_data = " << packed_qkv.GetByOffset("q_offset") << ";\n"
+                        << "  " << query.SetByIndices("vec3<u32>(batch_idx, seq_idx, hidden_idx)", "q_data") << ";\n"
```

Comment:
Since you are dispatching over Q domain, you can directly use `query.SetByOffset("global_idx", "q_data")`. And you can get `vec3<u32>(batch_idx, seq_idx, hidden_idx)` by `query.OffsetToIndices`  to replace above lines.